xds/src/main/java/io/grpc/xds/LocalityStore.java

```diff
@@ -149,23 +156,25 @@ public void handleSubchannelState(Subchannel subchannel, ConnectivityStateInfo n
 
     @Override
     public void reset() {
-      for (LocalityLbInfo localityLbInfo : localityMap.values()) {
-        localityLbInfo.shutdown();
+      for (XdsLocality locality : localityMap.keySet()) {
+        statsStore.removeLocality(locality);
+        localityMap.get(locality).shutdown();
       }
       localityMap = new HashMap<>();
     }
 
     // This is triggered by EDS response.
     @Override
-    public void updateLocalityStore(Map<Locality, LocalityInfo> localityInfoMap) {
-      Set<Locality> oldLocalities = localityMap.keySet();
-      Set<Locality> newLocalities = localityInfoMap.keySet();
+    public void updateLocalityStore(Map<XdsLocality, LocalityInfo> localityInfoMap) {
+      Set<XdsLocality> oldLocalities = localityMap.keySet();
+      Set<XdsLocality> newLocalities = localityInfoMap.keySet();
 
-      Iterator<Locality> iterator = oldLocalities.iterator();
+      Iterator<XdsLocality> iterator = oldLocalities.iterator();
       while (iterator.hasNext()) {
-        Locality oldLocality = iterator.next();
+        XdsLocality oldLocality = iterator.next();
         if (!newLocalities.contains(oldLocality)) {
           // No graceful transition until a high-level lb graceful transition design is available.
+          statsStore.removeLocality(oldLocality);
```

Comment:
A few requests that go to the removed localities before the picker gets updated will end up not being tracked. You need to do the removals after updating the picker.